File: Leetcode/res/Perfect Squares/2.cpp
\*
Author: allannozomu
Runtime: 96 ms
Memory: 11.4 MB*\

class Solution {
public:
    int numSquares(int n) {
        vector<int> dp = vector<int>(n + 1, 0);
        
        dp[1] = 1;
        for(int i = 2; i * i <=n; ++i){
            dp[i*i] = 1;
        }
        
        for (int i = 2; i <= n; ++i){
            if (dp[i]) continue;
            dp[i] = dp[i-1] + 1;
            for(int j = 2; j * j <=n; ++j){
                int perfect_square = j*j;
                if (i - perfect_square < 1) break;
                dp[i] = min(dp[i], dp[i - perfect_square]+1);
            }
        }
        
        return dp[n];
    }
};

File: Leetcode/res/Longest Common Prefix/2.py
# Author: allannozomu
# Runtime: 56 ms
# Memory: 13 MB

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        res = ""
        max_length = -1
        for s in strs:
            if max_length < 0:
                max_length = len(s)
            else:
                max_length = min(len(s), max_length)
       
        for i in range(max_length):
            c = ''
            for s in strs:
                if c == '':
                    c = s[i]
                elif c != s[i]:
                    return res
            res += c
        return res
            